src/MCP23017.cpp

```diff
@@ -7,40 +7,79 @@ MCP23017::MCP23017(uint8_t address, TwoWire& bus) {
 
 MCP23017::~MCP23017() {}
 
-void MCP23017::init()
-{
-	//BANK = 	0 : sequential register addresses
-	//MIRROR = 	0 : use configureInterrupt 
-	//SEQOP = 	1 : sequential operation disabled, address pointer does not increment
-	//DISSLW = 	0 : slew rate enabled
-	//HAEN = 	0 : hardware address pin is always enabled on 23017
-	//ODR = 	0 : open drain output
-	//INTPOL = 	0 : interrupt active low
+void MCP23017::begin(uint8_t pullup_a, uint8_t pullup_b)
+{
+	//IOCON.BANK   = 	0 : sequential register addresses
+	//IOCON.MIRROR = 	0 : use configureInterrupt 
+	//IOCON.SEQOP  = 	1 : sequential operation disabled, address pointer does not increment
+	//IOCON.DISSLW = 	0 : slew rate enabled
+	//IOCON.HAEN   = 	0 : hardware address pin is always enabled on 23017
+	//IOCON.ODR    = 	0 : open drain output
+	//IOCON.INTPOL = 	0 : interrupt active low
+	//IOCON bit 0 is spare
 	writeRegister(MCP23017_REGISTER::IOCON, 0b00100000);
 
 	//enable all pull up resistors (will be effective for input pins only)
-	writeRegister(MCP23017_REGISTER::GPPUA, 0xFF, 0xFF);
+	writeRegister(MCP23017_REGISTER::GPPUA, pullup_a, pullup_b);
 }
 
-void MCP23017::portMode(MCP23017_PORT port, uint8_t value)
+void MCP23017::portMode(MCP23017_PORT port, uint8_t modes, uint8_t pullups, uint8_t inverteds)
 {
-	writeRegister(MCP23017_REGISTER::IODIRA + port, value);
+   writeRegister(MCP23017_REGISTER::IODIRA + port, modes);
+   writeRegister(MCP23017_REGISTER::IPOLA  + port, inverteds);
+   writeRegister(MCP23017_REGISTER::GPPUA  + port, pullups);
 }
 
-void MCP23017::pinMode(uint8_t pin, uint8_t mode)
+void MCP23017::pinMode(uint8_t pin, uint8_t mode, bool inverted)
 {
 	MCP23017_REGISTER iodirreg = MCP23017_REGISTER::IODIRA;
-	uint8_t iodir;
+	MCP23017_REGISTER polreg   = MCP23017_REGISTER::IPOLA;
+	MCP23017_REGISTER pullreg  = MCP23017_REGISTER::GPPUA;
+
 	if(pin > 7)
 	{
 		iodirreg = MCP23017_REGISTER::IODIRB;
+		polreg = MCP23017_REGISTER::IPOLB;
+		pullreg = MCP23017_REGISTER::GPPUB;
 		pin -= 8;
 	}
 
-	iodir = readRegister(iodirreg);
-	if(mode == OUTPUT) iodir &= ~_BV(pin);
-	else iodir |= _BV(pin);
+	uint8_t iodir = readRegister(iodirreg);
+	uint8_t pol   = readRegister(polreg);
+	uint8_t pull  = readRegister(pullreg);
+
+	if(mode == OUTPUT) 
+	{
+		iodir |= _BV(pin);

```

Comment:
Why did you invert the iodir assignation logic between `OUTPUT` and `INPUT` ? An `OUTPUT` pin must be cleared according to the datasheet, so `iodir &= ~_BV(pin);` was the correct way of handling this.

Reply:
I did it first as you suggested, but the the result was just inverted as expected. This way solved that problem. My wiring was as described in the new example.

Reply:
I'm lost with that change. The code you're suggesting was initially the one I wrote, before it was corrected by #4.
I'll try to have a deeper look at this myself this weekend and sort it out once and for all. If that's true, then the datasheet is lying ^^

Reply:
So, I finally had the chance to get an mcp23017 out of the closet and test the library again.
I stick with my version of iodir, and this is very easy to prove :

I have a bunch of input switches on port B (active high), and some leds on port A. 

```cpp
#include <Wire.h>
#include <MCP23017.h>

#define MCP23017_ADDR 0x20
MCP23017 mcp = MCP23017(MCP23017_ADDR);

void setup() {
    Wire.begin();
    Serial.begin(115200);
    
    mcp.init();
    mcp.portMode(MCP23017_PORT::A, 0);         //Port A as output
    mcp.portMode(MCP23017_PORT::B, 0b11111111);//Port B as input

    mcp.writeRegister(MCP23017_REGISTER::GPIOA, 0x00); //Reset port A 
    mcp.writeRegister(MCP23017_REGISTER::GPIOB, 0x00); //Reset port B

    mcp.writeRegister(MCP23017_REGISTER::GPPUA, 0x00); //No pull ups on port A
    mcp.writeRegister(MCP23017_REGISTER::GPPUB, 0x00); //No pull ups on port B

    mcp.pinMode(7, INPUT);         //Pin 7 as input

    uint8_t conf = mcp.readRegister(MCP23017_REGISTER::IODIRA);
    Serial.print("IODIRA : ");
    Serial.print(conf, BIN);
    Serial.println();    
}


void loop() {
    uint8_t currentB;

    currentB = mcp.readPort(MCP23017_PORT::B);
    mcp.writePort(MCP23017_PORT::A, currentB);
}
```

> IODIRA : 10000000

In the example above, the led on pin 7 is not configured as an input, so it does not lit, whereas every other led do lit. This is with the current version of the library. Integrating your changes on `IODIR` gives

> IODIRA : 0

And all the leds are working (which proves that pin 7 is not configured as an input.

Reply:
Solved by v2.0.0. Feel free to reopen the discussion if you still disagree on the resolution. But in this is case, please read the comments on your example first. Your sketch has mistakes that clearly indicates you misused the library :)

Thanks for the suggestion, glad to see it finally made it to the library ! 👍